va/va.go

```diff
@@ -136,11 +137,48 @@ func (va ValidationAuthorityImpl) getAddr(ctx context.Context, hostname string)
 
 type dialer struct {
 	record core.ValidationRecord
+	stats  metrics.Scope
 }
 
 func (d *dialer) Dial(_, _ string) (net.Conn, error) {
 	realDialer := net.Dialer{Timeout: validationTimeout}
-	return realDialer.Dial("tcp", net.JoinHostPort(d.record.AddressUsed.String(), d.record.Port))
+	conn, err := realDialer.Dial("tcp", net.JoinHostPort(d.record.AddressUsed.String(), d.record.Port))
+
+	// If there was an error connecting to the preferred address and the IPv6First
+	// feature is enabled then attempt another connection if there is a fallback
+	// address available to try
+	if err != nil && features.Enabled(features.IPv6First) {
+		if fallback := fallbackAddress(d.record); fallback != nil {
+			d.record.AddressesTried = append(d.record.AddressesTried, d.record.AddressUsed)
+			d.record.AddressUsed = fallback
+			d.stats.Inc("IPv4Fallback", 1)
+			conn, err = realDialer.Dial("tcp", net.JoinHostPort(fallback.String(), d.record.Port))
+		}
+	}
+
+	return conn, err
+}
+
+// fallbackAddress takes a ValidationRecord and if the `AddressUsed` was IPv6 it
```

Comment:
This would remove the dependency on ordering within LookupHost. I think it would also simplify the logic in the calling function somewhat:

- split the slice
- if the feature flag is off:
  - append(IPv4, IPv6...), and try the first one
- if the feature flag is on:
  - try the first IPv6 address, if present
  - if timeout, or IPv6 not present, try the first IPv4 address